### DISCUSSION TITLE
Define arbitrary ray

### DISCUSSION THREAD

**Discussion by BDD-SDS** [mjhoptics/ray-optics#113] [General]
Hi @mjhoptics 

First of all thank you very much for providing this software!
I was wondering if it is possible to define custom rays? For this particular example I had an array of light sources in mind, where the light is sent to different directions using a lens system, see the figure below. I had hoped it is possible to define the starting point and angle of a couple of rays to define each light source and let the ray tracer calculate the overall path of every ray. Is this possible?

![ray tracing example](https://user-images.githubusercontent.com/128072813/225687087-41e7f7c3-b904-4569-9777-17095090e881.png)

Thank you very much!

**Comment by BDD-SDS** [comment]
I think I've figured it out using the trace module.

First the definition of the optical stack
```
def transform_ray_seg(poly, r, tfrm):
    rot, trns = tfrm
    p = rot.dot(r.p) + trns
    poly.append([p[2], p[1]])

opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']

osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=80)
osp['wvls'] = WvlSpec([(940.0, 1.0)], ref_wl=0)

opm.system_spec.dimensions = 'mm'
opm.radius_mode = True

sm.gaps[0].thi=20

sm.add_surface([1e4, 10, 'N-BK7','Schott'])
sm.set_stop()
sm.add_surface([-100, 100])

opm.update_model()
sm.list_model()
```

And then the raytracing
```
plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=False, do_draw_rays=False, do_draw_beams=False, do_draw_edge_rays=False).plot()

for m, pos in enumerate([-10, 0, 10]):
    for angle in [-10, -5, 0, 5, 10]:
        ray, opd, wvl = rt.trace(sm, pt0=np.array([0,pos,0]), dir0=np.array([0,np.sin(np.deg2rad(angle)), np.cos(np.deg2rad(angle))]), wvl=940.0)
        ray = [RaySeg(*rs) for rs in ray]

        path = []
        for i, r in enumerate(ray):
            transform_ray_seg(path, r, sm.gbl_tfrms[i])
        path = np.array(path)

        plt.plot(path[:,0], path[:,1], f'C{m}')
```

This gives the figure below
![image](https://user-images.githubusercontent.com/128072813/225960646-ec571901-915a-4b80-95a0-5786c8fa3bc7.png)

It's really a great tool once you've found your way!

**Comment by mjhoptics** [comment]
Hello @quark159,

Sorry for the delay in responding. Glad you found a solution with ray-optics. There are a variety of ways of doing this type of thing with ray-optics. Here's a road map of sorts to show how the pieces can be put together. The main parts are in the [raytr](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.html) package.

- [raytrace](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.raytrace.html) module: this contains the low level ray trace implementation. [trace_raw](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.raytrace.html#rayoptics.raytr.raytrace.trace_raw) can be called directly, if needed. 
- [trace](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.trace.html) module: This is the next level up in terms of ray trace interfaces. Ray specification is done via aperture and field coordinates. This is useful for analyzing imaging systems, but less so for more general inquiries. The [trace_base](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.trace.html#rayoptics.raytr.trace.trace_base) function computes the ray starting point and direction, given a relative pupil coordinate and field point, and calls `rt.trace`. The [trace_safe](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.trace.html#rayoptics.raytr.trace.trace_safe) function is a wrapper around `trace_base` that provides output control and options for handling ray error exceptions.
- [analyses](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.analyses.html) module: These build higher level functionality onto the individual ray tracing provided by the `raytrace` and `trace` modules. There are classes that manage tracing fans, lists and grids of rays. The specification is in terms of aperture and field (and wavelength) for the most part. There is a function, [trace_list_of_rays](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.raytr.analyses.html#rayoptics.raytr.analyses.trace_list_of_rays), that takes an optical model argument and a list of starting point, starting direction and wavelength, and returns the resulting rays.

These modules are used by the [layout](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.layout.html) and [interactivefigure](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.mpl.interactivefigure.html) subclasses to draw the lens pictures with the rays. They are also used by the graphics classes in [analysisfigure](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.mpl.analysisfigure.html) module as the data sources.

Thank you for posting the question and your solution; let me know if you have any other questions.

Regards,

Mike
